### PR TITLE
Global rework. Step 2: Gamification basics.

### PR DESCRIPTION
**Pull request by mrgambal**
ADD: Gamification is a simple blueprint within the application.
ADD: Global conceptions of rules, rule parsers and query builders.
ADD: Tests and documentation configs.
ADD: Basic Rule parsing validations.
ADD: Basic mongoDB query executor added.
ADD: New `RuleModel` class to reflect DB-specific model. ID is being generated as a hash of raw JSON-string being passed when loading a rule;

CHG: As constructor parameters list is hard to cope with, use named parameters notation to evade typos and parameter ordering mistakes.

### CODE REVIEW

File: vulyk/blueprints/gamification/core/rules.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        self._days_number = days_number
+        self._is_weekend = is_weekend
+        self._is_adjacent = is_adjacent
+
+        self._hash = id
+
+        self._validate()
+
+    def _validate(self):
+        """
```
Comment (dchaplinsky) [inline vulyk/blueprints/gamification/core/rules.py]: Maybe one day you should change that into list of tiny helpers, which you can iterate:

```self._validators = [validate_adj, validate_whatever]

def _validate(self):
 for v in self._validators:....
```

File: vulyk/blueprints/gamification/core/rules.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    @property
+    def is_weekend(self) -> bool:
+        return self._is_weekend
+
+    @property
+    def is_adjacent(self) -> bool:
+        return self._is_adjacent
+
+    @property
+    def limit(self):
```
Comment (dchaplinsky) [inline vulyk/blueprints/gamification/core/rules.py]: Logic here is not described in docstring.

Also it's quite an obscure property, in different contexts it means different things.

File: vulyk/blueprints/gamification/core/parsing.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+"""
+All available parsers, that convert raw representation could be received from
+any external source, are and should be kept here.
+"""
+import ujson as json
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/parsing.py]: ujson>=1.35

File: vulyk/blueprints/gamification/core/rules.py (hunk 2 above)
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/rules.py]: Done.

File: vulyk/blueprints/gamification/core/rules.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        """
+        # no achievements for n tasks closed on m adjacent days/weekends
+        # might be extended later
+        if self._is_adjacent:
+            if self._tasks_number > 0:
+                raise RuleValidationException(
+                    'Counting number of tasks closed in adjacent days/weekends'
+                    ' is not supported.')
+            elif self._days_number == 0:
+
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/rules.py]: Forry, miffed fis one.

File: vulyk/blueprints/gamification/core/rules.py (hunk 1 above)
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/rules.py]: I have considered several implementations, though as long as it is a private method, its interface or even inner principles do not really matter.

File: vulyk/blueprints/gamification/core/rules.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+    - you've been working for m weekends
+    - you've been working for m days in a row
+    - you've been working for m weekends in a row
+
+    Also additional bonuses to be given after user gets the achievement are
+    included as well as achievement and its badge is.
+    """
+
+    def __init__(self,
+                 id: int,
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/rules.py]: Done.

File: vulyk/blueprints/gamification/models/rules.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+from mongoengine import StringField, IntField, BooleanField
+
+from ..core.rules import Rule, ProjectRule
+
+
+class RuleModel(Document):
+    """
+    Database-specific rule representation
+    """
+    id = StringField(required=True, unique=True)
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/models/rules.py]: Let me take a quick look at the list of rules you gave me.

wait... I don't have one!

File: vulyk/blueprints/gamification/core/queries.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+        :return: True if user stats comply to the rule
+        :rtype: bool
+        """
+
+        # there is no way to get the size of aggregated batch using pymongo
+        # except explicit cursor dereferencing.
+        query = MongoRuleQueryBuilder(rule).build_for(user_id)
+
+        return sum([1 for _ in collection.aggregate(query)]) >= rule.limit
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/queries.py]: Done.

File: vulyk/blueprints/gamification/core/queries.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+        :returns: List of actions to be done within the aggregation.
+        :rtype: list[dict]
+        """
+        filter_first = self._filter_first.copy()
+        filter_first['user'] = user_id
+
+        result = [{'$match': filter_first}]
+
+        [result.append({statement.key: statement.clause}) for statement in [
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/queries.py]: Done.

File: vulyk/blueprints/gamification/core/queries.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        raise NotImplementedError()
+
+
+class MongoRuleQueryBuilder(RuleQueryBuilder):
+    """
+    Implementation of RuleQueryBuilder, bound to MongoDB.
+    """
+    _Pair = namedtuple('Pair', ['key', 'clause'])
+
+    def __init__(self, rule: Rule) -> None:
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/queries.py]: W/o comments it is less than 40 LOC and, furthermore, it does a single atomic task.

File: vulyk/blueprints/gamification/core/parsing.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    pass
+
+
+class JsonRuleParser(RuleParser):
+    """
+    Basic JSON parser.
+    """
+
+    @staticmethod
+    def parse(json_string: str) -> Rule:
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/parsing.py]: MY BRAIN HURTS!

File: vulyk/blueprints/gamification/core/__init__.py [hunk 11]
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
+"""
+A cask full of wonders: all of rule parsing and gamification logic.
```
Comment (mrgambal) [inline vulyk/blueprints/gamification/core/__init__.py]: C'mon. It's about the whole package, not the pitiful file itself.

**Pull request merged**